# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/ipc/ParentProcessDocumentChannel.cpp
# Commit: 3922996e51fa
# Full Hash: 3922996e51fa5029451101414af28d5bd54026d3
# Author: Jean-Yves Avenard <jyavenard@mozilla.com>
# Date: 2020-06-12 14:56:55
# Regressor Bug: 1641737
# File Overlap Count: 1
# Description:
#   Bug 1641737 - P7. Make the DLL and PPDC promises use direct task dispatch. r=mattwoodrow
#   
#   We attempt to reduce the number of event loop iterations, which would bring us back to a similar behaviour to the pre-DocumentChannel days.
#   
#   This prevent some tests to have increase intermittent failures.
# ==============================================================================

diff -r bcbaea8b572e -r 3922996e51fa netwerk/ipc/ParentProcessDocumentChannel.cpp
--- a/netwerk/ipc/ParentProcessDocumentChannel.cpp	Fri Jun 12 05:56:54 2020 +0000
+++ b/netwerk/ipc/ParentProcessDocumentChannel.cpp	Fri Jun 12 05:57:02 2020 +0000
@@ -66,6 +66,9 @@
   mStreamFilterEndpoints = std::move(aStreamFilterEndpoints);
 
   RefPtr<RedirectToRealChannelPromise> p = mPromise.Ensure(__func__);
+  // We make the promise use direct task dispatch in order to reduce the number
+  // of event loops iterations.
+  mPromise.UseDirectTaskDispatch(__func__);
 
   nsresult rv =
       gHttpHandler->AsyncOnChannelRedirect(this, channel, aRedirectFlags);
@@ -176,7 +179,7 @@
                 ->Then(
                     GetCurrentThreadSerialEventTarget(), __func__,
                     [self](RedirectToRealChannelPromise::ResolveOrRejectValue&&
-                               aValue) {
+                               aValue) -> RefPtr<RedirectToRealChannelPromise> {
                       MOZ_ASSERT(aValue.IsResolve());
                       nsresult rv = aValue.ResolveValue();
                       if (NS_FAILED(rv)) {
@@ -186,8 +189,12 @@
                       self->mListener = nullptr;
                       self->mCallbacks = nullptr;
                       self->RemoveObserver();
-                      return RedirectToRealChannelPromise::
-                          CreateAndResolveOrReject(std::move(aValue), __func__);
+                      auto p =
+                          MakeRefPtr<RedirectToRealChannelPromise::Private>(
+                              __func__);
+                      p->UseDirectTaskDispatch(__func__);
+                      p->ResolveOrReject(std::move(aValue), __func__);
+                      return p;
                     });
         // We chain the promise the DLL is waiting on to the one returned by
         // RedirectToRealChannel. As soon as the promise returned is
